devdoc/umock_c_lib_requirements_with_ids.md

```diff
@@ -1202,3 +1202,87 @@ X**SRS_UMOCK_C_LIB_01_195: [** If any error occurs during the destroy_call relat
 XX**SRS_UMOCK_C_LIB_01_204: [** Tracking of paired calls shall not be done if the actual call to the `create_call` is using the `SetFailReturn` call modifier. **]**
+
+## real function support
+
+Sometimes it is desirable to track the mockable function calls and also call the real implementation of those functions.
+While that is possible by simply registering a hook function for each of the mockable functions, it is also rather involved and cumbersome.
+
+In order to simplify the process, the macros `MOCKABLE_INTERFACE`, `IMPLEMENT_MOCKABLE_FUNCTION` and `REGISTER_GLOBAL_INTERFACE_HOOKS` can be used.
+
+Example:
+
+Given a test dependency with the below interface:
+
+```c
+MOCKABLE_INTERFACE(test_interface,
+    FUNCTION(, int, test_dependency_1_arg, int, a),
+    FUNCTION(, int, test_dependency_2_args, int, a, int, b)
+)
+```
+
+A user would write in the unit test code:
+
+```c
+#define ENABLE_MOCKS
+
+#include "test_dependency.h"
+#include "test_dependency_real_code.c"
+
+TEST_FUNCTION(reals_are_setup_at_interface_level)
+{
+    // arrange
+    int result;
+
+    REGISTER_GLOBAL_INTERFACE_HOOKS(test_interface);
+
+    STRICT_EXPECTED_CALL(test_dependency_1_arg(45));
+
+    // act
+    result = test_dependency_1_arg(45);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, "", umock_c_get_expected_calls());
+    ASSERT_ARE_EQUAL(char_ptr, "", umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int, 42, result);
+}
+```
+
+The functions defined in `test_dependency_real_code.c` have to use `IMPLEMENT_MOCKABLE_FUNCTION`.
+The inclusion of the `test_dependency_real_code.c` file while `ENABLE_MOCKS` is defined ensures that the functions implemented in that file get automatically renamed to `real_` functions.
+
+### MOCKABLE_INTERFACE
+
+```c
+#define MOCKABLE_INTERFACE(interface_name, ...) \
+```
+
+`MOCKABLE_INTERFACE` allows grouping several mockable functions in a header in order to be able to register all the hooks into one call for the entire group of functions.
+
+XX**SRS_UMOCK_C_LIB_01_215: [** Each item in `...` shall be an entry for one mockable function. **]**
+
+XX**SRS_UMOCK_C_LIB_01_216: [** Each item in `...` shall be defined using a macro called `FUNCTION`, which shall be an alias for `MOCKABLE_FUNCTION`. **]**
+
+Example:
+
+```c
+MOCKABLE_INTERFACE(test_interface,
+    FUNCTION(, int, test_dependency_1_arg, int, a),
+    FUNCTION(, int, test_dependency_2_args, int, a, int, b)
+)
+```
+
+### IMPLEMENT_MOCKABLE_FUNCTION
+
+```c
+#define IMPLEMENT_MOCKABLE_FUNCTION(modifiers, result, function, ...) \
+    ...
+```
+
+XX**SRS_UMOCK_C_LIB_01_217: [** In the presence of the `ENABLE_MOCKS` define, `IMPLEMENT_MOCKABLE_FUNCTION` shall expand to the signature of the function, but the name shall be changed to be prefix with `real_`. **]**
+
+XX**SRS_UMOCK_C_LIB_01_218: [** If `ENABLE_MOCKS` is not defined, `IMPLEMENT_MOCKABLE_FUNCTION` shall expand to the signature of the function. **]**
+
+### REGISTER_GLOBAL_INTERFACE_HOOKS
```

Comment:
>REGISTER_GLOBAL_INTERFACE_HOOKS [](start = 4, length = 31)

missing triple backtick c and the definition of the macro. Like line 1257 for example.

Reply:
Added

---
In reply to: [293554873](https://github.com/Azure/umock-c/pull/95#discussion_r293554873) [](ancestors = 293554873)